File: Part1_강의자료/Ch07_이분탐색/문제별코드/2295_세수의합/2295.py
N = int(input())
arr = sorted([int(input()) for _ in range(N)])
sums = sorted([(arr[i] + arr[j]) for i in range(N) for j in range(i, N)])

def is_exist(A):
    left, right = 0, len(sums)
    while left <= right:
        mid = (left + right) // 2
        if sums[mid] < A:
            left = mid + 1
        elif sums[mid] > A:
            right = mid - 1
        else:
            return 1
    return 0

flag = False
for i in range(N - 1, 0, -1):
    for j in range(i - 1, -1, -1):
        if is_exist(arr[i] - arr[j]):
            print(arr[i])
            flag = True
            break
    if flag is True:
        break

File: Part1_강의자료/Ch07_이분탐색/문제별코드/2295_세수의합/2295.cpp
#include <algorithm>
#include <stdio.h>
#include <vector>
using namespace std;

int is_exist(int N, vector<int> &vec, int A)
{
    int left = 0, right = N - 1, mid;

    while (left <= right)
    {
        mid = (left + right) / 2;
        if (vec[mid] < A)
            left = mid + 1;
        else if (vec[mid] > A)
            right = mid - 1;
        else
            return 1;
    }
    return 0;
}

int main()
{
    int N;

    scanf("%d", &N);
    vector<int> vec(N);
    for (int i = 0; i < N; i++)
        scanf("%d", &vec[i]);

    vector<int> sums(N * (N + 1) / 2);
    int sum_index = 0;
    for (int i = 0; i < N; i++)
        for (int j = i; j < N; j++)
            sums[sum_index++] = vec[i] + vec[j];

    sort(vec.begin(), vec.end());
    sort(sums.begin(), sums.end());
    for (int i = N - 1; i > 0; i--)
    {
        for (int j = i - 1; j >= 0; j--)
        {
            if (is_exist(sum_index, sums, vec[i] - vec[j]))
            {
                printf("%d\n", vec[i]);
                return 0;
            }
        }
    }

    return 0;
}
